numpy/random/_common.pyx

```diff
@@ -404,6 +417,7 @@ cdef object cont_broadcast_1(void *func, void *state, object size, object lock,
     randoms_data = <double *>np.PyArray_DATA(randoms)
     n = np.PyArray_SIZE(randoms)
     it = np.PyArray_MultiIterNew2(randoms, a_arr)
+    validate_output_shape(it.shape, randoms)
```

Comment:
@seberg, is this something that `PyArray_MultiIterNew2` ought to handle for output arrays? Or would we need to use a different type of iterator instead?

Reply:
Not sure if this answers your question as to why this is here, but here goes.

It handles the case where they don't broadcast.  The issue here happens when the iterator is say (m,n) but randoms is (m,1). This results in the output (which is randoms) being repeatedly filled and only containing the last value where it came up from the iterator.  

Essentially randoms has to be (a) broadcastable with the broadcast size of the broadcast value of all of the functions inputs and (b) weakly larger than the broadcast size all inputs. Say if there are 3 inputs, a,b,c, then we can define input_shape=(a+b+c).shape. If `n=len(input_shape)`, then it much be the case that

```
for i_s, o_s in zip(input_shape, output_shape[-n:]):
    assert i_s == o_s or (o_s > i_s and i_s==1)
```

This was a manual check in the old RandomState.

Reply:
My claim is simply that I'm surprised our iteration constructors don't already have checks for such a case.

Reply:
Well, to be fair `PyArray_MultiIterNew` which this uses is the old iterator API. The new iterator API will have larger overheads for small arrays, but otherwise be much faster (ufuncs have an "all inputs contiguous or 1D" special case because of that.

So yes, I think this is expected. The old style iterator does simple broadcasting. The new-style could be flagged to only allow `out` to be larger, but not smaller (at this time).

Reply:
Should we be converting all of the random stuff to use the new iterator?